pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go

```diff
@@ -305,37 +331,48 @@ func (s *Container) getStatsForStmt(
 	// We first try and see if we can get by without creating a new entry for this
 	// key, as this allows us to not construct the statementFingerprintID from scratch (which
 	// is an expensive operation)
-	stats, _, _ = s.getStatsForStmtWithKey(key, invalidStmtFingerprintID, false /* createIfNonexistent */)
+	stats, _, _ = s.getStatsForStmtWithKey(key, invalidStmtFingerprintID, false /* createIfNonexistent */, false /* unsafe */)
 	if stats == nil {
 		stmtFingerprintID = constructStatementFingerprintIDFromStmtKey(key)
-		stats, created, throttled = s.getStatsForStmtWithKey(key, stmtFingerprintID, createIfNonexistent)
+		stats, created, throttled = s.getStatsForStmtWithKey(key, stmtFingerprintID, createIfNonexistent, false /* unsafe */)
 		return stats, key, stmtFingerprintID, created, throttled
 	}
 	return stats, key, stats.ID, false /* created */, false /* throttled */
 }
 
+// getStatsForStmtWithKey returns an instance of stmtStats.
+// If createIfNonexistent flag is set to true, then a new entry is created in
+// the Container if it does not yet exist.
+// If unsafe flag is set to true, then the new entry is fetched/created without
+// locking onto the mutex. This should only be used when caller is absolutely
+// sure that the Container is not being used concurrently by multiple goroutines.
```

Comment:
Rather than calling it unsafe, what if we called it `doNotLock` or `alreadyLocked`. It might be the case that it's not explicitly locked above this call but it might be implicitly locked. It's not actually "unsafe" to not lock it, it just means that that locking is being handled at a different level.

Alternatively, much of the rest of the cockroach codebase would split this into two methods, one called `getStatsForStmtWithKey` and another called `getStatsForStmtWithKeyLocked` where the first is:

```go
func (s *Container) getStatsForStmtWithKey(
	key stmtKey, stmtFingerprintID roachpb.StmtFingerprintID, createIfNonexistent bool, unsafe bool,
) (stats *stmtStats, created, throttled bool) {
    s.mu.Lock()
    defer s.mu.Unlock()
    return  s.getStatsForStmtWithKey(key, stmtFingerprintID, createIfNonexistent)
}
```